paddlenlp/transformers/ctrl/modeling.py

```diff
@@ -0,0 +1,764 @@
+# Copyright 2018 Salesforce and HuggingFace Inc. team.
+# Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" paddle2.x CTRL model."""
+
+import numpy as np
+import paddle
+import paddle.nn as nn
+import paddle.nn.functional as F
+from paddle.nn import CrossEntropyLoss, MSELoss
+from .. import PretrainedModel, register_base_model
+
+from .utils import (
+    BaseModelOutputWithPast,
+    CausalLMOutputWithPast,
+    Config,
+    SequenceClassifierOutput, )
+
+__all__ = [
+    'CTRLModel', "CTRLLMHeadModel", 'CTRLForSequenceClassification',
+    'SinusoidalPositionalEmbedding'
+]
+
+
+class SinusoidalPositionalEmbedding(nn.Embedding):
+    def __init__(self, num_embeddings, embedding_dim):
+        super().__init__(num_embeddings, embedding_dim)
+        self.weight = self._init_weight(self.weight)
+
+    @staticmethod
+    def _init_weight(out):
+        n_pos, dim = out.shape
+        out.stop_gradient = True
+        position_ids = paddle.arange(0, n_pos, dtype=out.dtype).unsqueeze(1)
+        indices = paddle.arange(0, dim // 2, dtype=out.dtype).unsqueeze(0)
+
+        indices = 10000.0**(-2 * indices / dim)
+        embeddings = paddle.matmul(position_ids, indices)
+        sentinel = dim // 2
+        out[:, 0:sentinel] = paddle.sin(embeddings)
+        out[:, sentinel:] = paddle.cos(embeddings)
+
+        return out
+
+    @paddle.no_grad()
+    def forward(self, position_ids):
+        return super().forward(position_ids)
+
+
+def scaled_dot_product_attention(q, k, v, mask, attention_mask=None):
+    # calculate attention
+    matmul_qk = paddle.matmul(q, k, transpose_y=True)
+
+    scaled_attention_logits = matmul_qk / np.sqrt(k.shape[-1])
+
+    if mask is not None:
+        nd, ns = scaled_attention_logits.shape[
+            -2], scaled_attention_logits.shape[-1]
+        scaled_attention_logits += mask[ns - nd:ns, :ns] * -1e4
+
+    if attention_mask is not None:
+        # Apply the attention mask
+        scaled_attention_logits = scaled_attention_logits + attention_mask
+
+    attention_weights = F.softmax(scaled_attention_logits, axis=-1)
+
+    output = paddle.matmul(attention_weights, v)
+
+    return output, attention_weights
+
+
+class MultiHeadAttention(nn.Layer):
+    def __init__(self, d_model_size, num_heads):
+        super().__init__()
+        self.num_heads = num_heads
+        self.d_model_size = d_model_size
```

Comment:
Better to use hidden_size instead of d_model_size

Reply:
done